lib/std/zon/parse.zig

```diff
@@ -0,0 +1,2832 @@
+                    free(self.gpa, item);
+                }
+            };
+
+            result[i] = try self.parseExpr(array_info.child, options, nodes.at(@intCast(i)));
+        }
+        return result;
+    }
+
+    fn parseStruct(
+        self: *@This(),
+        comptime T: type,
+        comptime options: Options,
+        node: Zoir.Node.Index,
+    ) !T {
+        const repr = node.get(self.zoir);
+        const fields: @FieldType(Zoir.Node, "struct_literal") = switch (repr) {
+            .struct_literal => |nodes| nodes,
+            .empty_literal => .{ .names = &.{}, .vals = .{ .start = node, .len = 0 } },
+            else => return self.failExpectedContainer(T, node),
+        };
+
+        const field_infos = @typeInfo(T).@"struct".fields;
+
+        // Gather info on the fields
+        const field_indices = b: {
+            comptime var kvs_list: [field_infos.len]struct { []const u8, usize } = undefined;
+            inline for (field_infos, 0..) |field, i| {
+                kvs_list[i] = .{ field.name, i };
+            }
+            break :b std.StaticStringMap(usize).initComptime(kvs_list);
+        };
+
+        // Parse the struct
+        var result: T = undefined;
+        var field_found: [field_infos.len]bool = .{false} ** field_infos.len;
+
+        // If we fail partway through, free all already initialized fields
+        var initialized: usize = 0;
+        errdefer if (options.free_on_error and field_infos.len > 0) {
+            for (fields.names[0..initialized]) |name_runtime| {
+                switch (field_indices.get(name_runtime.get(self.zoir)) orelse continue) {
+                    inline 0...(field_infos.len - 1) => |name_index| {
+                        const name = field_infos[name_index].name;
+                        free(self.gpa, @field(result, name));
+                    },
+                    else => unreachable, // Can't be out of bounds
+                }
+            }
+        };
+
+        // Fill in the fields we found
+        for (0..fields.names.len) |i| {
+            const field_index = b: {
+                const name = fields.names[i].get(self.zoir);
+                break :b field_indices.get(name) orelse if (options.ignore_unknown_fields) {
+                    continue;
+                } else {
+                    return self.failUnexpectedField(T, node, i);
+                };
+            };
+
+            // We now know the array is not zero sized (assert this so the code compiles)
+            if (field_found.len == 0) unreachable;
+
+            if (field_found[field_index]) {
+                var buf: [2]Ast.Node.Index = undefined;
+                const struct_init = self.ast.fullStructInit(&buf, node.getAstNode(self.zoir)).?;
+                const field_node = struct_init.ast.fields[i];
+                const token = self.ast.firstToken(field_node) - 2;
+                return self.failTokenFmt(token, 0, "duplicate struct field name", .{});
+            }
```

Comment:
Oh, this is a good point, ZonGen should be doing this check if it's not already (AstGen does the equivalent check for Zig).

Reply:
ZonGen isn't doing it right now, but I assumed that was intentional--

I basically get this check "for free" here since I have to look up each string regardless to fill out the data. OTOH, ZonGen would have to duplicate these efforts if it wanted to detect this error.

Maybe ZON doesn't actually forbid duplicate fields, it's just that Zig does, and therefore it's up to the code that lowers ZON to Zig to enforce this rule, not the code that parses the ZON.

Reply:
That last question is the one that matters. I'm not entirely sure of the answer, but my instinct is that ZON should disallow duplicate fields:
* It's more in alignment with Zig
* I don't see any clear use cases for duplicate fields
* It allows more errors to be caught at the file level, which is better for `@import` of ZON

Reply:
I gave this some more thought and you're right. My motivation for avoiding doing the check in ZonGen was to avoid nlogn string comparisons when it could be done with n boolean comparisons later. But then I remembered that ZonGen interns strings anyway so it's just nlogn integer comparisons which I'm not concerned about.

I went ahead and implemented duplicate field checking in ZonGen, and removed it from parse and LowerZon. This also let me simplify the error handling in LowerZon a bit, this was the last holdout that couldn't report an error without converting the location from a Zoir node to an Ast tyupe first.